### PR TITLE
Implement proper touch screen support by emulating the stylus

### PR DESCRIPTION
**Pull request by jroweboy**
Video of the PR in action: https://youtu.be/7Y-X4Eg7FiA

Using a mouse for touch screen controls doesn't accurately encompass the
feel that one gets when using the 3ds stylus. This PR adds experimental
support for using a stylus for touch screen controls.

Controls are:
 * Left Clicking on the pointer half of the stylus and dragging rotates the stylus. 
 * Left Clicking on the back half (telescoping end) and dragging moves the stylus.
 * Double Clicking the stylus touches it to the screen.
 * Double Click and drag functions holds the stylus to the screen and moves it with the mouse.

### CODE REVIEW

File: src/citra_qt/bootmanager.cpp [hunk 1]
```diff
@@ -209,30 +225,67 @@ void GRenderWindow::keyReleaseEvent(QKeyEvent* event) {
...
+    } else if (event->button() == Qt::RightButton) {
+        motion_emu->Tilt(pos.x(), pos.y());
+    }
+}
+void GRenderWindow::mouseDoubleClickEvent(QMouseEvent* e) {
+    if (e->button() == Qt::LeftButton && stylus->Contains(e->pos())) {
+        qreal pixelRatio = windowPixelRatio();
+        QPoint p = stylus->GetTouchPoint();
+        QPoint screen = mapToGlobal(QWidget::geometry().topLeft());
+        LOG_ERROR(Frontend, "screen %d, %d", screen.x(), screen.y());
```
Comment (JayFoxRox) [inline src/citra_qt/bootmanager.cpp]: This shouldn't be an error log entry?

File: src/citra_qt/bootmanager.h [hunk 2]
```diff
@@ -127,12 +130,14 @@ class GRenderWindow : public QWidget, public EmuWindow {
     void mousePressEvent(QMouseEvent* event) override;
     void mouseMoveEvent(QMouseEvent* event) override;
     void mouseReleaseEvent(QMouseEvent* event) override;
+    void GRenderWindow::mouseDoubleClickEvent(QMouseEvent* e) override;
```
Comment (JayFoxRox) [inline src/citra_qt/bootmanager.h]: Contains class name

File: src/citra_qt/bootmanager.h [hunk 3]
```diff
@@ -127,12 +130,14 @@ class GRenderWindow : public QWidget, public EmuWindow {
...
 
     void focusOutEvent(QFocusEvent* event) override;
 
     void OnClientAreaResized(unsigned width, unsigned height);
 
     void InitRenderTarget();
+    Stylus* GRenderWindow::GetStylus();
```
Comment (JayFoxRox) [inline src/citra_qt/bootmanager.h]: ...

File: src/citra_qt/citra-qt.rc [hunk 4]
```diff
@@ -6,4 +6,3 @@
 // Icon with lowest ID value placed first to ensure application icon
 // remains consistent on all systems.
 CITRA_ICON              ICON                    "../../dist/citra.ico"
-
```
Comment (JayFoxRox) [inline src/citra_qt/citra-qt.rc]: unintentional change?

File: src/citra_qt/main.cpp [hunk 5]
```diff
@@ -71,10 +71,13 @@ GMainWindow::GMainWindow() : config(new Config()), emu_thread(nullptr) {
 
     setWindowTitle(QString("Citra %1| %2-%3")
                        .arg(Common::g_build_name, Common::g_scm_branch, Common::g_scm_desc));
+    setWindowIcon(QIcon(":citra_icon.ico"));
```
Comment (JayFoxRox) [inline src/citra_qt/main.cpp]: I think this should be in another PR or at least another commit

File: src/citra_qt/stylus.cpp [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    qreal x = n.x() - center.x();
+    qreal y = n.y() - center.y();
+
+    qreal mv_angle = std::atan2(y, x);
+
+    // get the changed angle
+    rotation = (mv_angle - initial_angle) * 180 / M_PI;
+
+    if (std::fabs(rotation) > 360.0) {
+        rotation = fmod(rotation, 360);
```
Comment (JayFoxRox) [inline src/citra_qt/stylus.cpp]: `360.0` ?

File: src/citra_qt/stylus.cpp [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    }
+}
+
+bool Stylus::IsNearCenter(QPoint n) {
+    // QPoint c = GetCenter();
+    // return (n - c).manhattanLength() < 75;
```
Comment (JayFoxRox) [inline src/citra_qt/stylus.cpp]: Dead code

File: src/citra_qt/stylus.cpp [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    // jokes on you i'm now checking to see if its the top half or the bottom half
```
Comment (JayFoxRox) [inline src/citra_qt/stylus.cpp]: I think this is a good and logical design decision. No need for the comment

**Comment by JayFoxRox** [comment]
I'll be honest: I think the implementation is stupid. The code quality is not up to par.

However, I do like the approach you are taking. If we want to get Citra VR-ready we will also need a stable stylus emulation (within a 3D world).
But at this point, the lack of control over the stylus makes this PR pointless. You should have used a 3D stylus as rotating the stylus in only 1 axes is not enough. I also want to angle it.

Your current implementation also lacks more direct feedback: When double clicking the stylus it should show an animation.
For a more direct mapping between real world and the virtual stylus I'd also recommend considering to control the stylus with 2 mice at once (one controlling upper part, the other one the lower tip). If you are set on that 1 mouse approach we should at least raise / lower the virtual stylus with the mousewheel so we can get rid of the stupid double-click.
(This might require a virtual shadow to be able to tell if the stylus is on the display or not, but this could be a follow-up PR)

I also think it should be extended by finger-scratching or the ability to load custom models / images as object to touch the display with. I'm not saying I have a fetish for licking screens for touch input, but I'm guessing some people would want such features. Again: not me, just asking for a friend.

---

That being said: It's fine as a temporary solution to a permanent problem (better stylus emulation).
Please fix some of the remaining issues so we can have this in bleeding edge and unofficial builds sooner :)

File: src/citra_qt/overlay.h [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+#include <QPainter>
+#include <QtWidgets>
+
+class Stylus;
+class GRenderWindow;
+
+class Overlay : public QWidget {
+    Q_OBJECT
+public:
+    Overlay(GRenderWindow* parent);
```
Comment (lioncash) [inline src/citra_qt/overlay.h]: `explicit Overlay(GRenderWindowParent* parent)`

File: src/citra_qt/overlay.h [hunk 10]
```diff
@@ -0,0 +1,32 @@
+// Copyright 20XX Citra Emulator Project
+// Licensed under GPLv2 or any later version (but not GPLv824 cause that one sucked)
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <QPainter>
+#include <QtWidgets>
```
Comment (lioncash) [inline src/citra_qt/overlay.h]: Please don't use `<QtWidgets>`. Explicitly include the widget headers themselves.

File: src/citra_qt/stylus.cpp [hunk 11]
```diff
@@ -0,0 +1,84 @@
+#include <QApplication>
+#include <QtWidgets>
```
Comment (lioncash) [inline src/citra_qt/stylus.cpp]: Ditto

File: src/citra_qt/stylus.cpp [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+#include "common/logging/log.h"
+
+Stylus::Stylus() {
+    pic = QPixmap(":stylus.png");
+    x = QApplication::desktop()->screenGeometry().width() / 2;
+    y = QApplication::desktop()->screenGeometry().height() / 2;
+    rotation = 0;
+}
+
+Stylus::~Stylus() {}
```
Comment (lioncash) [inline src/citra_qt/stylus.cpp]: `Stylus::~Stylus() = default;`

File: src/citra_qt/stylus.cpp [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+#include <QtWidgets>
+
+#include "citra_qt/stylus.h"
+#include "common/logging/log.h"
+
+Stylus::Stylus() {
+    pic = QPixmap(":stylus.png");
+    x = QApplication::desktop()->screenGeometry().width() / 2;
+    y = QApplication::desktop()->screenGeometry().height() / 2;
+    rotation = 0;
```
Comment (lioncash) [inline src/citra_qt/stylus.cpp]: This can be initialized directly in the class body.

File: src/citra_qt/stylus.cpp [hunk 14]
```diff
@@ -0,0 +1,84 @@
+#include <QApplication>
+#include <QtWidgets>
+
+#include "citra_qt/stylus.h"
+#include "common/logging/log.h"
+
+Stylus::Stylus() {
+    pic = QPixmap(":stylus.png");
```
Comment (lioncash) [inline src/citra_qt/stylus.cpp]: These can be directly initialized in the constructor initializer list.

File: src/citra_qt/stylus.h [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+#include <QPainter>
+#include <QVector2D>
+
+enum class StylusState : unsigned { ROTATE, HOLD, DROP, COUNT };
+
+class Stylus {
+public:
+    Stylus();
+    ~Stylus();
+    StylusState GetState() {
```
Comment (lioncash) [inline src/citra_qt/stylus.h]: Please place all implementations in the cpp file. This can also be a const member function.

File: src/citra_qt/stylus.h [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+
+    void Rotate(QPoint p);
+
+    void SetHoldPoint(QPoint p);
+
+    void SetState(StylusState s) {
+        state = s;
+    }
+    void UpdatePosition(unsigned x, unsigned y);
+    unsigned GetX() {
```
Comment (lioncash) [inline src/citra_qt/stylus.h]: Const member function

File: src/citra_qt/stylus.h [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+        return x;
+    }
+    unsigned GetY() {
```
Comment (lioncash) [inline src/citra_qt/stylus.h]: Const member function

File: src/citra_qt/stylus.h [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+        return y;
+    }
+
+    QPixmap GetPix() {
```
Comment (lioncash) [inline src/citra_qt/stylus.h]: Const member function.

File: src/citra_qt/stylus.h [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+        return pic;
+    }
+
+    QTransform GetTransform();
```
Comment (lioncash) [inline src/citra_qt/stylus.h]: Const member function.

File: src/citra_qt/stylus.h [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+
+    QPolygon GetPoly();
```
Comment (lioncash) [inline src/citra_qt/stylus.h]: Const member function. Also `GetPolygon()` is more consistent with the other getters.

File: src/citra_qt/stylus.h [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+
+    bool Contains(QPoint p);
```
Comment (lioncash) [inline src/citra_qt/stylus.h]: Const member function.

File: src/citra_qt/stylus.h [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+
+    QPoint GetTouchPoint();
```
Comment (lioncash) [inline src/citra_qt/stylus.h]: Const member function.

File: src/citra_qt/stylus.h [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+
+    bool IsNearCenter(QPoint p);
```
Comment (lioncash) [inline src/citra_qt/stylus.h]: Const member function.

File: src/citra_qt/stylus.h [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+
+    QPoint GetCenter();
```
Comment (lioncash) [inline src/citra_qt/stylus.h]: Const member function.

File: src/citra_qt/stylus.h [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+
+    QRect GetRect();
```
Comment (lioncash) [inline src/citra_qt/stylus.h]: Const member function.

File: src/citra_qt/overlay.cpp [hunk 26]
```diff
@@ -0,0 +1,66 @@
+
+#include <QPainter>
+
+#include "citra_qt/bootmanager.h"
+#include "citra_qt/overlay.h"
+#include "citra_qt/stylus.h"
+
+Overlay::Overlay(GRenderWindow* parent) : QWidget(parent), parent(parent) {
+    // setWindowFlags(Qt::Widget | Qt::FramelessWindowHint);
```
Comment (lioncash) [inline src/citra_qt/overlay.cpp]: These should likely not be left in the code.

File: src/citra_qt/overlay.cpp [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+    parent->mouseReleaseEvent(event);
+    // window()->setFocus();
+    QApplication::setActiveWindow(window());
+}
+
+void Overlay::paintEvent(QPaintEvent* ev) {
+    Stylus* stylus = parent->GetStylus();
+    QPainter painter(this);
+    // set the viewport to match the screen size
+    // painter.setWindow(QApplication::desktop()->screenGeometry());
```
Comment (lioncash) [inline src/citra_qt/overlay.cpp]: Ditto about this commented out code.

File: src/citra_qt/overlay.cpp [hunk 28]
```diff
@@ -0,0 +1,66 @@
...
+    parent->keyReleaseEvent(event);
+}
+
+void Overlay::mousePressEvent(QMouseEvent* event) {
+    parent->mousePressEvent(event);
+    QApplication::setActiveWindow(window());
+}
+void Overlay::mouseMoveEvent(QMouseEvent* event) {
+    parent->mouseMoveEvent(event);
+    // window()->setFocus();
```
Comment (lioncash) [inline src/citra_qt/overlay.cpp]: Ditto about this commented out code.

File: src/citra_qt/overlay.cpp [hunk 29]
```diff
@@ -0,0 +1,66 @@
...
+    QApplication::setActiveWindow(window());
+}
+void Overlay::mouseDoubleClickEvent(QMouseEvent* event) {
+    parent->mouseDoubleClickEvent(event);
+    // window()->setFocus();
```
Comment (lioncash) [inline src/citra_qt/overlay.cpp]: Ditto about this commented out code.

File: src/citra_qt/overlay.cpp [hunk 30]
```diff
@@ -0,0 +1,66 @@
...
+    QApplication::setActiveWindow(window());
+}
+void Overlay::mouseReleaseEvent(QMouseEvent* event) {
+    parent->mouseReleaseEvent(event);
+    // window()->setFocus();
```
Comment (lioncash) [inline src/citra_qt/overlay.cpp]: Ditto about this commented out code.

File: src/citra_qt/stylus.cpp [hunk 31]
```diff
@@ -0,0 +1,84 @@
...
+    x = QApplication::desktop()->screenGeometry().width() / 2;
+    y = QApplication::desktop()->screenGeometry().height() / 2;
+    rotation = 0;
+}
+
+Stylus::~Stylus() {}
+
+void Stylus::SetHoldPoint(QPoint p) {
+    // hold = QPoint(p.x() - x, p.y() - y);
```
Comment (lioncash) [inline src/citra_qt/stylus.cpp]: Is this intentionally left in?

File: src/citra_qt/overlay.cpp [hunk 32]
```diff
@@ -0,0 +1,66 @@
...
+    // setAttribute(Qt::WA_NoSystemBackground, true);
+    // setAttribute(Qt::WA_TranslucentBackground, true);
+    // setWindowFlags(Qt::FramelessWindowHint | Qt::ToolTip | Qt::WindowStaysOnTopHint);
+    setWindowFlags(Qt::FramelessWindowHint | Qt::ToolTip);
+    setAttribute(Qt::WA_TransparentForMouseEvents);
+    setAttribute(Qt::WA_NoSystemBackground);
+    setAttribute(Qt::WA_TranslucentBackground);
+    setFocusPolicy(Qt::NoFocus);
+    setMouseTracking(false);
+    move(0, 0);
```
Comment (lioncash) [inline src/citra_qt/overlay.cpp]: Why is this necessary?

File: src/citra_qt/bootmanager.cpp [hunk 33]
```diff
@@ -209,30 +225,67 @@ void GRenderWindow::keyReleaseEvent(QKeyEvent* event) {
...
 }
 
 void GRenderWindow::mouseMoveEvent(QMouseEvent* event) {
     auto pos = event->pos();
-    qreal pixelRatio = windowPixelRatio();
-    this->TouchMoved(std::max(static_cast<unsigned>(pos.x() * pixelRatio), 0u),
-                     std::max(static_cast<unsigned>(pos.y() * pixelRatio), 0u));
-    motion_emu->Tilt(pos.x(), pos.y());
+
+    if (stylus->GetState() == StylusState::HOLD) {
```
Comment (lioncash) [inline src/citra_qt/bootmanager.cpp]: This if-else should be within its own function

File: src/citra_qt/bootmanager.cpp [hunk 34]
```diff
@@ -209,30 +225,67 @@ void GRenderWindow::keyReleaseEvent(QKeyEvent* event) {
...
+        qreal pixelRatio = windowPixelRatio();
+        QPoint p = stylus->GetTouchPoint();
+        QPoint screen = mapToGlobal(QWidget::geometry().topLeft());
+        this->TouchMoved(std::max(static_cast<unsigned>((p.x() - screen.x()) * pixelRatio), 0u),
+                         std::max(static_cast<unsigned>((p.y() - screen.y()) * pixelRatio), 0u));
+    } else if (stylus->GetState() == StylusState::ROTATE) {
+        stylus->Rotate(pos);
+        overlay->update();
+    }
+    if (event->button() == Qt::LeftButton) {
```
Comment (lioncash) [inline src/citra_qt/bootmanager.cpp]: Ditto for this code. Also an empty if-statement looks kind of wonky.

File: src/citra_qt/bootmanager.cpp [hunk 35]
```diff
@@ -148,6 +153,16 @@ void GRenderWindow::DoneCurrent() {
 
 void GRenderWindow::PollEvents() {}
 
+void GRenderWindow::UpdateStylusSize(unsigned width, unsigned height) {
+    if (overlay) {
```
Comment (lioncash) [inline src/citra_qt/bootmanager.cpp]:
```cpp
if (!overlay)
  return;
```

File: src/citra_qt/bootmanager.h [hunk 36]
```diff
@@ -9,10 +9,13 @@
 #include <mutex>
 #include <QGLWidget>
 #include <QThread>
+#include "citra_qt/overlay.h"
+#include "citra_qt/stylus.h"
 #include "common/thread.h"
 #include "core/frontend/emu_window.h"
 #include "core/frontend/motion_emu.h"
 
+
```
Comment (lioncash) [inline src/citra_qt/bootmanager.h]: Unnecessary newline?

File: src/citra_qt/bootmanager.h (hunk 3 above)
Comment (lioncash) [inline src/citra_qt/bootmanager.h]: This can also be a const member function.

**Comment by v1tal** [comment]
i cant believe you guys picked apart this obvious joke so much

**Comment by Luffykun007** [comment]
I think there are more important things to implement or fix than the need for a styllus in the emulator right now but it's just my thinking.

**Comment by wwylele** [comment]
What's the status of this PR?

**Comment by RavenHome1** [comment]
@wwylele april fools is over so..

**Comment by jfmherokiller** [comment]
this looks pretty good i wonder if it will allow me to use my wacom tablet as the touch screen

**Comment by JayFoxRox** [comment]
YES BE NOW!

**Comment by bunnei** [comment]
It's clear from the feedback that this change is missing a lot of critical features to the user experience. The author furthermore has not addressed any of this feedback, so I'm marking this as abandoned and closing it.
Feel free to submit a new PR with the feedback addressed. Thanks!
